gk/main.c

```diff
@@ -157,28 +158,72 @@ priority_from_delta_time(uint64_t present, uint64_t past)
 	return integer_log_base_2(delta_time);
 }
 
+static int
+get_grantor_id(struct gk_lpm *ltbl, struct ip_flow *flow)
+{
+	int ret;
+	if (flow->proto == ETHER_TYPE_IPv4)
+		ret = lpm_lookup_ipv4(ltbl->lpm, flow->f.v4.dst);
+	else
+		ret = lpm_lookup_ipv6(ltbl->lpm6, flow->f.v6.dst);
+
+	return ret;
+}
+
+static struct gk_policy *
+get_gk_policy(struct gk_lpm *ltbl, struct ip_flow *flow)
+{
+	int gt_id = get_grantor_id(ltbl, flow);
+	return gt_id < 0 ? NULL : &ltbl->policy_tbl[gt_id];
+}
+
 static inline void
-initialize_flow_entry(struct flow_entry *fe, struct ip_flow *flow)
+initialize_flow_entry(struct flow_entry *fe,
+	struct ip_flow *flow, struct gk_lpm *ltbl)
 {
 	rte_memcpy(&fe->flow, flow, sizeof(*flow));
 
 	fe->state = GK_REQUEST;
 	fe->u.request.last_packet_seen_at = rte_rdtsc();
 	fe->u.request.last_priority = START_PRIORITY;
 	fe->u.request.allowance = START_ALLOWANCE - 1;
-	/* TODO Grantor ID comes from LPM lookup. */
-	fe->u.request.grantor_id = 0;
+
+	/* Grantor ID comes from LPM lookup. */
+	if (fe->u.request.grantor_id == 0) {
+		int gt_id = get_grantor_id(ltbl, &fe->flow);
+		if (gt_id < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: initialize flow entry error in %s\n",
+				__func__);
+			return;
+		}
+
+		fe->u.request.grantor_id = gt_id;
+	}
 }
 
 static inline void
-reinitialize_flow_entry(struct flow_entry *fe, uint64_t now)
+reinitialize_flow_entry(
+	struct flow_entry *fe, struct gk_lpm *ltbl, uint64_t now)
 {
+	int gt_id;
+
 	fe->state = GK_REQUEST;
 	fe->u.request.last_packet_seen_at = now;
 	fe->u.request.last_priority = START_PRIORITY;
 	fe->u.request.allowance = START_ALLOWANCE - 1;
-	/* TODO Grantor ID comes from LPM lookup. */
-	fe->u.request.grantor_id = 0;
+
+	/* Grantor ID comes from LPM lookup. */
+	gt_id = get_grantor_id(ltbl, &fe->flow);
```

Comment:
If `fe` is at the state `GK_REQUEST` or `GK_GRANTED` when `reinitialize_flow_entry()` is called, you're supposed to save `grantor_id` to avoid another lookup. Moreover, notice that `grantor_id` leaves in a `union`, so you need to copy it before resetting the fields.